File: tests/evidence.ts
import type { Page, TestInfo } from '@playwright/test';

function fileSafe(label: string): string {
  return label.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '') || 'step';
}

/** Perform one user action, verify its post-condition, and attach evidence of that state. */
export async function actionEvidence(
  page: Page,
  testInfo: TestInfo,
  label: string,
  action: () => Promise<void>,
  verify: () => Promise<void>,
) {
  if (typeof action !== 'function' || typeof verify !== 'function') throw new TypeError('actionEvidence requires action and verify callbacks');
  await action();
  if (typeof verify !== 'function') throw new TypeError('observationEvidence requires a verify callback');
  await verify();
  const path = testInfo.outputPath(`${fileSafe(label)}.png`);
  await page.screenshot({ path, fullPage: true });
  await testInfo.attach(label, { path, contentType: 'image/png' });
}

export async function observationEvidence(
  page: Page,
  testInfo: TestInfo,
  label: string,
  verify: () => Promise<void>,
) {
  if (typeof verify !== 'function') throw new TypeError('observationEvidence requires a verify callback');
  await verify();
  const path = testInfo.outputPath(`${fileSafe(label)}.png`);
  await page.screenshot({ path, fullPage: true });
  await testInfo.attach(label, { path, contentType: 'image/png' });
}

File: website_test_pipeline/evidence.py
from pathlib import Path
import re

def action_evidence(page, test_name: str, action, verify, directory: Path) -> Path:
    action(); verify()
    directory.mkdir(parents=True, exist_ok=True)
    safe_name = re.sub(r"[^A-Za-z0-9_.-]+", "-", test_name).strip("-") or "evidence"
    path = directory / f"{safe_name}.png"
    page.screenshot(path=str(path), full_page=True)
    return path

def observation_evidence(page, test_name: str, verify, directory: Path) -> Path:
    verify(); return action_evidence(page, test_name, lambda: None, lambda: None, directory)
